include/drivers/rtio_sensor.h

```diff
@@ -0,0 +1,127 @@
+#define ZEPHYR_INCLUDE_DRIVERS_RTIO_SENSOR_H_
+/**
+ * @brief RTIO Sensor Interface
+ * @defgroup rtio_sensor_interface RTIO Sensor Interface
+ * @ingroup io_interfaces
+ * @{
+ */
+
+#include <kernel.h>
+#include <device.h>
+#include <drivers/rtio.h>
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+/**
+ * @file
+ * @brief Real-Time IO Sensor API
+ *
+ * Each  RTIO Sensor acts as a pollable stream of buffers with a driver that
+ * provides a way of deserializing the raw sensor data buffers into meaningful
+ * values.
+ *
+ * Each RTIO Sensor is expected to output buffers containing sample sets (such
+ * as 9 axis IMU data) in a device specific format understood by the driver.
+ * Each driver must then implement the sensor reader api and functionality
+ * to deserialize this stream of buffers into meaningful SI unit data. The
+ * caller may request different numerical formats of this data.
+ *
+ * *TODO* To start only 32bit floating point numerical formatted values are
+ * provided. Q format values with scale factors would be very useful for
+ * use with CMSIS DSP or many SoCs that do not contain FPUs. The reader
+ * *should* provide simple ways of converting integer readings into
+ * useful numerical formats in common functions.
+ */
+
+/**
+ * @private
+ */
+struct rtio_sensor_reader;
+
+/**
+ * @private
+ * @brief Get the number of sample sets in the current block
+ */
+typedef ssize_t (*rtio_sensor_reader_sample_sets_t)(const struct sensor_reader *reader);
+
+/**
+ * @private
+ * @brief Iterate to the next sample set in the block
+ *
+ * Returns 0 when complete, 1 otherwise
+ *
+ */
+typedef int (*rtio_sensor_reader_next_t)(struct rtio_sensor_reader *reader);
+
+/**
+ * @private
+ * @brief Get the value of a sample sets channel, if it exists
+ *
+ * The channel is an index into a device specified sample channel.
+ * For example, A 3 axis accelerometer may assign channel 0 to
+ * the X axis, 1 to the Y axis, and 2 to the Z axis.
+ *
+ * *TODO* discuss how the channel identifier might be obtained for something like
+ * the X channel so that the same API calls may be used for multiple devices.
+ */
+typedef int (*rtio_sensor_reader_get_float_t)(const struct sensor_reader *reader,
+					      const u32_t channel, float *val);
+
+
+/**
+ * @private
+ * @brief Calculate or return the timestamp of the current sample set
+ *
+ * The timestamp may be part of the sample set or may be interpolated from the
+ * blocks begin and end timestamps along with the number of sample sets in the
+ * block or it may be in the block data itself.
+ */
+typedef int (*rtio_sensor_reader_timestamp_t)(const struct sensor_reader *reader,
+					      u32_t *timestamp);
+
+/**
+ * @brief RTIO Sensor Reader
+ */
+struct rtio_sensor_reader {
+	rtio_sensor_sample_sets_t sample_sets;
+	rtio_sensor_reader_next_t next;
+	rtio_sensor_reader_get_float_t get_float;
+	rtio_sensor_reader_timestamp_t timestamp;
+
+	struct rtio_block *block;
+	size_t position;
```

Comment:
I'm confused about how this would all work together with the underlying RTIO block. Maybe it's because I haven't looked carefully enough at the test case for RTIO, though last time I checked it didn't exercise all the API.

I'm guessing that the control flow is something like this in a basic, do-it-all-from-the-main-thread application:

```
int main()
{
	setup();
	while (true) {
		sleep_until_block_has_data_ready();
		while (sample_sets())
		{
			get_float();
			do_something_with_the_float();
			next();
		}
	}
}
```

Reply:
The general usage would be something like...

``` c
int main() {
   struct k_fifo *myfifo = // some valid fifo here
   struct rtio_allocator *myalloc = // some valid allocator here
   struct device *mysensor = // some valid instance of rtio_sensor here
  
   struct rtio_configuration config =  { .fifo = myfifo, .allocator = myallocator, // more settings as needed, assume trigger is called by gpio or timer };
   rtio_configure(mydev, &config);

   // our sensor is now getting data and storing it in allocated buffers automatically, we need to poll or it will start failing to allocate soon-ish

   struct rtio_sensor_reader reader;
   while(true) {
      struct rtio_block *myblock = k_fifo_get(myfifo, K_FOREVER);
      rtio_sensor_reader(mydev, myblock, &myreader);
      while(rtio_sensor_reader_next(&myreader)) {
            u32_t timestamp;
           float acc_x, acc_y, acc_z;
          rtio_sensor_reader_get_float(&myreader, ACC_X, &acc_x);
          rtio_sensor_reader_get_float(&myreader, ACC_Y, &acc_y);
          rtio_sensor_reader_get_float(&myreader, ACC_Z, &acc_z);
          rtio_sensor_reader_timestamp(&myreader, &timestamp);
          printf("cycle: %d, acc (m/s^2) x,y,z: %f, %f, %f\n", timestamp, acc_x, acc_y, acc_z);
      }
   }
}
```